Tests/SwiftPrometheusTests/SwiftPrometheusTests.swift

```diff
@@ -40,76 +40,162 @@ final class SwiftPrometheusTests: XCTestCase {
             self.myValue = myValue
         }
     }
-
     
-    var prom: Prometheus!
+    
+    var prom: PrometheusClient!
     
     override func setUp() {
-        self.prom = Prometheus()
+        self.prom = PrometheusClient()
     }
     
     override func tearDown() {
         self.prom = nil
     }
     
     func testCounter() {
+        let semaphore = DispatchSemaphore(value: 0)
```

Comment:
For the entire file:
You don't need your `semaphore.wait()` after every `inc()`, `dec()` or other method that changes the value of the metric, since the library takes care of thread safety and the `getMetric` will not run while there are still other tasks (observes) in the queue.

Reply:
Just for me to follow along right, that's because the queue has `.barrier` set, right?

Reply:
Hm, I'm not so sure this is true; looks like we do need to syncronize, maybe?

```
Test Suite 'All tests' started at 2019-04-07 10:10:48.126
Test Suite 'SwiftPrometheusTests.xctest' started at 2019-04-07 10:10:48.127
Test Suite 'SwiftPrometheusTests' started at 2019-04-07 10:10:48.128
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testCounter]' started.
/Users/jmsmith/workspace/github.com/MrLotU/SwiftPrometheus/Tests/SwiftPrometheusTests/SwiftPrometheusTests.swift:67: error: -[SwiftPrometheusTests.SwiftPrometheusTests testCounter] : XCTAssertEqual failed: ("10") is not equal to ("20")
2019-04-07 10:10:48.413898-0700 atos[98193:28783184] examining /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Xcode/Agents/xctest [98185]
==================
WARNING: ThreadSanitizer: Swift access race (pid=98185)
  Modifying access of Swift variable at 0x7b180000ad30 by thread T4:
    #0 Counter.metrics.modify <compiler-generated> (Prometheus:x86_64+0x2235)
    #1 closure #1 in Counter.inc(_:_:_:) Counter.swift:75 (Prometheus:x86_64+0x61e4)
    #2 partial apply for closure #1 in Counter.inc(_:_:_:) <compiler-generated> (Prometheus:x86_64+0x66c7)
    #3 thunk for @escaping @callee_guaranteed () -> () <compiler-generated> (Prometheus:x86_64+0x4da0)
    #4 __tsan::invoke_and_release_block(void*) <null>:2135536 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x6a42b)
    #5 _dispatch_client_callout <null>:2135536 (libdispatch.dylib:x86_64+0x3dce)
    #6 _dispatch_client_callout <null>:2135536 (libdispatch.dylib:x86_64+0x3dce)
    #7 _dispatch_client_callout <null>:2135536 (libdispatch.dylib:x86_64+0x3dce)

  Previous read of size 8 at 0x7b180000ad30 by main thread:
    #0 Counter.metrics.getter <compiler-generated> (Prometheus:x86_64+0x20a0)
    #1 Counter.get(_:) Counter.swift:92 (Prometheus:x86_64+0x6bb4)
    #2 implicit closure #3 in SwiftPrometheusTests.testCounter() <compiler-generated> (SwiftPrometheusTests:x86_64+0xb262)
    #3 partial apply for implicit closure #3 in SwiftPrometheusTests.testCounter() <compiler-generated> (SwiftPrometheusTests:x86_64+0xb2de)
    #4 thunk for @callee_guaranteed () -> (@unowned Int, @error @owned Error) <compiler-generated> (SwiftPrometheusTests:x86_64+0xa25d)
    #5 thunk for @callee_guaranteed () -> (@unowned Int, @error @owned Error)partial apply <compiler-generated> (SwiftPrometheusTests:x86_64+0xb39e)
    #6 closure #1 in XCTAssertEqual<A>(_:_:_:file:line:) <null>:2135536 (libswiftXCTest.dylib:x86_64+0xb063)
    #7 @objc SwiftPrometheusTests.testCounter() <compiler-generated> (SwiftPrometheusTests:x86_64+0xbae0)
    #8 __invoking___ <null>:2135536 (CoreFoundation:x86_64h+0x4af1b)

  Location is heap block of size 88 at 0x7b180000ace0 allocated by main thread:
    #0 malloc <null>:2135568 (libclang_rt.tsan_osx_dynamic.dylib:x86_64h+0x4caea)
    #1 swift_slowAlloc <null>:2135568 (libswiftCore.dylib:x86_64+0x2c91f8)
    #2 PrometheusClient.createCounter<A, B>(forType:named:helpText:initialValue:withLabelType:) Prometheus.swift:50 (Prometheus:x86_64+0x2f95f)
    #3 SwiftPrometheusTests.testCounter() SwiftPrometheusTests.swift:58 (SwiftPrometheusTests:x86_64+0x95e0)
    #4 @objc SwiftPrometheusTests.testCounter() <compiler-generated> (SwiftPrometheusTests:x86_64+0xbae0)
    #5 __invoking___ <null>:2135568 (CoreFoundation:x86_64h+0x4af1b)

  Thread T4 (tid=28783161, running) is a GCD worker thread

SUMMARY: ThreadSanitizer: Swift access race <compiler-generated> in Counter.metrics.modify
==================
ThreadSanitizer report breakpoint hit. Use 'thread info -s' to get extended information about the report.
Program ended with exit code: 9
```

Reply:
I think that race was because I was trying to `get` the value from inside the closure I passed to the `DispatchQueue`. Here's where I'm seeing not everything is being written out unless I `wait()` and `signal()` each time:

```
Test Suite 'All tests' started at 2019-04-07 10:39:29.832
Test Suite 'SwiftPrometheusTests.xctest' started at 2019-04-07 10:39:29.833
Test Suite 'SwiftPrometheusTests' started at 2019-04-07 10:39:29.834
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testCounter]' started.
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testCounter]' passed (0.115 seconds).
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testGauge]' started.
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testGauge]' passed (0.002 seconds).
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testHistogram]' started.
/Users/jmsmith/workspace/github.com/MrLotU/SwiftPrometheus/Tests/SwiftPrometheusTests/SwiftPrometheusTests.swift:132: error: -[SwiftPrometheusTests.SwiftPrometheusTests testHistogram] : XCTAssertEqual failed: ("# HELP my_histogram Histogram for testing
# TYPE my_histogram histogram
my_histogram_bucket{myValue="*", le="0.5"} 0.0
my_histogram_bucket{myValue="*", le="1.0"} 0.0
my_histogram_bucket{myValue="*", le="2.0"} 0.0
my_histogram_bucket{myValue="*", le="3.0"} 0.0
my_histogram_bucket{myValue="*", le="5.0"} 0.0
my_histogram_bucket{myValue="*", le="+Inf"} 0.0
my_histogram_count{myValue="*"} 0.0
my_histogram_sum{myValue="*"} 0.0
my_histogram_bucket{myValue="labels", le="0.5"} 0.0
my_histogram_bucket{myValue="labels", le="1.0"} 0.0
my_histogram_bucket{myValue="labels", le="2.0"} 0.0
my_histogram_bucket{myValue="labels", le="3.0"} 0.0
my_histogram_bucket{myValue="labels", le="5.0"} 0.0
my_histogram_bucket{myValue="labels", le="+Inf"} 0.0
my_histogram_count{myValue="labels"} 0.0
my_histogram_sum{myValue="labels"} 0.0") is not equal to ("# HELP my_histogram Histogram for testing
# TYPE my_histogram histogram
my_histogram_bucket{myValue="*", le="0.5"} 0.0
my_histogram_bucket{myValue="*", le="1.0"} 1.0
my_histogram_bucket{myValue="*", le="2.0"} 2.0
my_histogram_bucket{myValue="*", le="3.0"} 4.0
my_histogram_bucket{myValue="*", le="5.0"} 4.0
my_histogram_bucket{myValue="*", le="+Inf"} 4.0
my_histogram_count{myValue="*"} 4.0
my_histogram_sum{myValue="*"} 9.0
my_histogram_bucket{myValue="labels", le="0.5"} 0.0
my_histogram_bucket{myValue="labels", le="1.0"} 0.0
my_histogram_bucket{myValue="labels", le="2.0"} 0.0
my_histogram_bucket{myValue="labels", le="3.0"} 1.0
my_histogram_bucket{myValue="labels", le="5.0"} 1.0
my_histogram_bucket{myValue="labels", le="+Inf"} 1.0
my_histogram_count{myValue="labels"} 1.0
my_histogram_sum{myValue="labels"} 3.0")
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testHistogram]' failed (0.010 seconds).
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testInfo]' started.
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testInfo]' passed (0.001 seconds).
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testSummary]' started.
/Users/jmsmith/workspace/github.com/MrLotU/SwiftPrometheus/Tests/SwiftPrometheusTests/SwiftPrometheusTests.swift:163: error: -[SwiftPrometheusTests.SwiftPrometheusTests testSummary] : XCTAssertEqual failed: ("# HELP my_summary Summary for testing
# TYPE my_summary summary
my_summary{quantile="0.5", myValue="*"} 4.0
my_summary{quantile="0.9", myValue="*"} 10000.0
my_summary{quantile="0.99", myValue="*"} 10000.0
my_summary_count{myValue="*"} 1.0
my_summary_sum{myValue="*"} 0.0
my_summary{quantile="0.5", myValue="labels"} 0.0
my_summary{quantile="0.9", myValue="labels"} 0.0
my_summary{quantile="0.99", myValue="labels"} 0.0
my_summary_count{myValue="labels"} 0.0
my_summary_sum{myValue="labels"} 0.0") is not equal to ("# HELP my_summary Summary for testing
# TYPE my_summary summary
my_summary{quantile="0.5", myValue="*"} 4.0
my_summary{quantile="0.9", myValue="*"} 10000.0
my_summary{quantile="0.99", myValue="*"} 10000.0
my_summary_count{myValue="*"} 5.0
my_summary_sum{myValue="*"} 10130.0
my_summary{quantile="0.5", myValue="labels"} 123.0
my_summary{quantile="0.9", myValue="labels"} 123.0
my_summary{quantile="0.99", myValue="labels"} 123.0
my_summary_count{myValue="labels"} 1.0
my_summary_sum{myValue="labels"} 123.0")
Test Case '-[SwiftPrometheusTests.SwiftPrometheusTests testSummary]' failed (0.011 seconds).
Test Suite 'SwiftPrometheusTests' failed at 2019-04-07 10:39:29.977.
	 Executed 5 tests, with 2 failures (0 unexpected) in 0.139 (0.143) seconds
Test Suite 'SwiftPrometheusTests.xctest' failed at 2019-04-07 10:39:29.978.
	 Executed 5 tests, with 2 failures (0 unexpected) in 0.139 (0.145) seconds
Test Suite 'All tests' failed at 2019-04-07 10:39:29.980.
	 Executed 5 tests, with 2 failures (0 unexpected) in 0.139 (0.148) seconds
Program ended with exit code: 1
```

Reply:
Ah, this is from the `.observe()` we updated [below](https://github.com/MrLotU/SwiftPrometheus/pull/2#discussion_r272791913).

```
if let labels = labels ... {
    let his = ...
    his.observe(value) // This is run *async* and thus may not be completed by the time the `done` callback hanlder is called!!!
}
self.total.inc(value)

for (i, bound) in ... {
    ...
}
done()
```

Reply:
Open to feedback on a better way to fix this, perhaps this suggestion [from the commit message](https://github.com/MrLotU/SwiftPrometheus/pull/2/commits/dbbf9505cdf361e519d96d76a542a8bd25133cef)?

Reply:
See my comment on said function in the Histogram implementation :)

Reply:
I don't exactly remember the thought process behind this comment 😂 

Anyhow, you can leave this in any way you'd like right now, since the rewrite using the apple/swift-metrics library will use locks instead of callbacks to keep things thread safe, but will not need all the semaphore & other async workarounds.

Reply:
Awesome! Sounds great